example/locked_cout.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright Takatoshi Kondo 2019
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(MQTT_LOCKED_COUT_HPP)
+#define MQTT_LOCKED_COUT_HPP
+
+#include <mutex>
+#include <iostream>
+
+class locked_stream {
+public:
+    locked_stream(std::ostream& stream)
+        :lock_(mtx_),
+         stream_(stream) {}
+
+    friend
+    locked_stream&& operator<<(locked_stream&& s, std::ostream& (*arg)(std::ostream&)) {
+        s.stream_ << arg;
+        return std::move(s);
+    }
+
+    template <typename Arg>
+    friend
+    locked_stream&& operator<<(locked_stream&& s, Arg&& arg) {
+        s.stream_ << std::forward<Arg>(arg);
+        return std::move(s);
+    }
+
+private:
+    static std::mutex mtx_;
+    std::unique_lock<std::mutex> lock_;
+    std::ostream& stream_;
+
+};
```

Comment:
I've solved this problem this way in the past:

```cpp
struct locked
{
    std::ostream& cout() const { return std::cout; }
    std::ostream& cerr() const { return std::cerr; }

private:
    static std::mutex& mutex()
    {
        static std::mutex stdio_mutex;
        return stdio_mutex;
    }
    std::unique_lock<std::mutex> lock_{mutex()};
};

// Report a failure
void
fail(std::error_code ec)
{
    locked().cerr() << ec.message() << std::endl;
}
```

Reply:
How does it work? I can't image client code of locked.

Reply:
Just like this:

```cpp
    locked().cerr() << ec.message() << std::endl;
```
You construct a new instance of it when you want to write, and call either cerr() or cout() to get the stream that you're writing to.

I'm not saying you should change your code to do this. I was simply showing you an example.

You might consider using the function-static variable for the mutex though, since it keeps the initialization of the mutex variable in one place.